test/WebJobs.Script.Tests/BlobLeaseManagerTests.cs

```diff
@@ -0,0 +1,316 @@
+                await blob.ReleaseLeaseAsync(new AccessCondition { LeaseId = leaseId });
+            }
+
+            resetEvent.Wait(TimeSpan.FromSeconds(15));
+
+            manager.Dispose();
+
+            Assert.True(resetEvent.IsSet);
+            Assert.True(manager.HasLease, $"{nameof(BlobLeaseManager.HasLease)} was not correctly set to 'true' when lease was acquired.");
+            Assert.Equal(hostId, manager.LeaseId);
+        }
+
+        [Fact]
+        [Trait("Category", "E2E")]
+        public async Task HasLeaseChanged_WhenLeaseIsLostAndStateChanges_IsFired()
+        {
+            string connectionString = AmbientConnectionStringProvider.Instance.GetConnectionString(ConnectionStringNames.Storage);
+            ICloudBlob blob = await GetLockBlobAsync(connectionString);
+
+            string hostId = Guid.NewGuid().ToString();
+            var traceWriter = new TestTraceWriter(TraceLevel.Verbose);
+            var resetEvent = new ManualResetEventSlim();
+
+            BlobLeaseManager manager = null;
+            string tempLeaseId = null;
+
+            using (manager = new BlobLeaseManager(blob, TimeSpan.FromSeconds(15), hostId, traceWriter, TimeSpan.FromSeconds(3)))
+            {
+                try
+                {
+                    await TestHelpers.Await(() => manager.HasLease);
+
+                    manager.HasLeaseChanged += (s, a) => resetEvent.Set();
+
+                    // Release the manager's lease and acquire one with a different id
+                    await blob.ReleaseLeaseAsync(new AccessCondition { LeaseId = manager.LeaseId });
+                    tempLeaseId = await blob.AcquireLeaseAsync(TimeSpan.FromSeconds(30), Guid.NewGuid().ToString());
+                }
+                finally
+                {
+                    if (tempLeaseId != null)
+                    {
+                        await blob.ReleaseLeaseAsync(new AccessCondition { LeaseId = tempLeaseId });
+                    }
+                }
+
+                resetEvent.Wait(TimeSpan.FromSeconds(15));
+            }
+
+            Assert.True(resetEvent.IsSet);
+            Assert.False(manager.HasLease, $"{nameof(BlobLeaseManager.HasLease)} was not correctly set to 'false' when lease lost.");
+        }
+
+        [Fact]
+        public void AcquiringLease_WithServerError_LogsAndRetries()
+        {
+            var hostId = "123";
+            var traceWriter = new TestTraceWriter(TraceLevel.Verbose);
+            var renewResetEvent = new ManualResetEventSlim();
+
+            var results = new Queue<Task<string>>();
+            results.Enqueue(Task.FromException<string>(new StorageException(new RequestResult { HttpStatusCode = 500 }, "test", null)));
+            results.Enqueue(Task.FromResult(hostId));
+            
+            var blobMock = new Mock<ICloudBlob>();
+            blobMock.Setup(b => b.AcquireLeaseAsync(It.IsAny<TimeSpan>(), It.IsAny<string>()))
+                .Returns(() => results.Dequeue());
```

Comment:
I did go down the path of using `SetupSequence` (and would prefer to use that) but changed it because it lacks a convenient mechanism to enable the logic we're performing in the callback to signal the `ManualResetEventSlim`. I might have missed something, though, so I'm open to ideas.